sdcm/cluster.py

```diff
@@ -2434,111 +2443,226 @@ def check_node_health(self):
+    def search_for_nulls_in_peers(self, gossip_info, peers_details):
+        """
+        This validation is added to recreate the issue: https://github.com/scylladb/scylla/issues/4652
+        Found scenario described in the https://github.com/scylladb/scylla/issues/6397
+        """
+        if not gossip_info:
+            self.log.warning("Gossip info is not available. Search for the warning above")
+            return
 
-        # Validate that JUST existent nodes in the gossip
-        not_in_cluster = list(set(gossip_node_schemas.keys()) - set(cluster_nodes))
-        if not_in_cluster:
-            ClusterHealthValidatorEvent(type='error', name='GossipNodeSchemaVersion', status=Severity.ERROR,
-                                        node=self.name,
-                                        error='Node(s) %s exists in the gossip, but doesn\'t exist in the cluster'
-                                              % ', '.join(ip for ip in not_in_cluster))
-            for ip in not_in_cluster:
-                gossip_node_schemas.pop(ip)
-
-        # Validate that same schema on all nodes
-        if len(set(node_info['schema'] for node_info in gossip_node_schemas.values())) > 1:
-            ClusterHealthValidatorEvent(type='error', name='GossipNodeSchemaVersion', status=Severity.ERROR,
-                                        node=self.name,
-                                        error='Schema version is not same on all nodes in gossip info: {}'
-                                        .format('\n'.join('{}: {}'.format(ip, schema_version['schema'])
-                                                          for ip, schema_version in gossip_node_schemas.items())))
+        for ip, node_info in peers_details.items():
+            nulls = [column for column, value in node_info.items() if value == 'null']
+            if nulls:
+                is_target = self.if_node_is_target_node(ip)
+                if ip in gossip_info and gossip_info[ip]['status'] not in self.GOSSIP_STATUSES_FILTER_OUT:
+                    ClusterHealthValidatorEvent(type='error', name='NodePeersNulls', status=Severity.ERROR,
+                                                node=self.name,
+                                                error=f"Current node {self.ip_address}. Found nulls in system.peers for "
+                                                      f"node {ip}{is_target} with status {gossip_info[ip]['status']} : "
+                                                      f"{peers_details[ip]}")
+                else:
+                    # Issue https://github.com/scylladb/scylla/issues/6397 - Should the info about decommissioned node
+                    # be kept in the system.peers?
+                    self.log.warning(f"Current node {self.ip_address}. Found nulls in system.peers for "
+                                     f"node {ip}{is_target} with status {gossip_info[ip]['status']} : {peers_details[ip]}")
```

Comment:
Looks like messages are the same. Do we need to inform that the node is decommissioned? Anyway it's better to format common parts of the messages ones.

Reply:
Fixed